### ISSUE TITLE
Blob hash mismatch (integrity check failed)

### ISSUE THREAD

**Issue by A139834** [Azure/aztk#551]
Hi
We have been using Spark (via `aztk`) to access blob storage (`wasbs://...`). Occasionally for very large instances (e.g. 64 cores) and very large binary files (i.e. 50GB parquet) we are seeing errors relating to MD5 mismatches:

```
Caused by: com.microsoft.azure.storage.StorageException: Blob hash mismatch (integrity check failed), Expected value is wGiPnU44QHHWmUTKtUrpXA==, retrieved jntR+kYPt3BMSSQiASXPKA==.
	at com.microsoft.azure.storage.blob.CloudBlob$9.postProcessResponse(CloudBlob.java:1549)
	at com.microsoft.azure.storage.blob.CloudBlob$9.postProcessResponse(CloudBlob.java:1452)
```

The read and write was both created with these verisons of the libraries `hadoop-azure-2.7.4.jar` and `azure-storage-3.1.0.jar`. Amazingly if we lower the paralleism by reducing the number of cores available to Spark the errors do not happen.

I have tested and found that disabling the md5 does allow the job to complete and the data does not appear to actually be corrupt (which I would expect would be a fairly major error on the parquet decompression if it actually was corrupt):

```
--conf spark.hadoop.fs.azure.check.block.md5=false
```

Has anyone else run in to this issue with blob/spark?

**Comment by A139834** [comment]
Hi,
Sorry I have tried all the hadoop-azure-2.7.x series drivers trying to debug this issue and they are all affected with this problem (up to 2.7.6).

I think I need to raise this against the `hadoop-azure` repo directly.

**Comment by jafreck** [comment]
I'm assuming you are using aztk version 0.7.0 (the latest version). If so, the new aztk Docker images (the ones you get when you specify the `toolkit` block in cluster.yaml) come with Hadoop 2.8.3. So the `hadoop-azure-2.8.3.jar` is already present on the image. So is the corresponding azure-storage client jar. 

With both of these present on the image out of the box, you don't need to add any jars to interface with Azure Storage. 

Can you try accessing wasb without adding any other jars to the cluster?

**Comment by jafreck** [comment]
@A139834 have you had any success with the above suggestions?

**Comment by A139834** [comment]
@jafreck Hi
Sorry I haven't had a chance yet. I should be able to test in the next 24-36 hours.

**Comment by A139834** [comment]
An update:

READING:

I have tested these azure-storage libraries with builds of Spark with Hadoop 2.7.3 (`hadoop-azure-2.7.6.jar`) and Hadoop 2.8.3 (`hadoop-azure-2.8.3.jar`):
azure-storage-2.2.0.jar
azure-storage-3.1.0.jar
azure-storage-4.0.0.jar

All have the same MD5 hash issue:
```
Blob hash mismatch (integrity check failed), Expected value is kvz2/qYj8qCHytF9l5PjjA==, retrieved /7205Nf5+1ieAU3NqdMaHQ==.
```
but the azure-storage-2.2.0.jar has a slightly better error message:
```
[masked].../part-00050-8d1cc5e4-bc0e-4beb-a9bb-c9885d6902f8-c000.snappy.parquet Exception details: java.io.IOException Error Code : InvalidMd5
```

If I incrementally update to the latest version of azure-storage with either of the Hadoop versions:
azure-storage-5.0.0.jar -> blows up `com.microsoft.azure.storage.core.Utility.initIOException(Utility.java:737)`

I am able to successfully read the same file which failed reliably above with both Hadoop 2.7.3 and 2.8.3 (albeit with weird 12 minute timeout delays which i think are rate limiting):
azure-storage-6.0.0.jar
azure-storage-7.0.0.jar

However that leads to the next problem:

WRITING:

You can see in the release notes for the azure-storage library (https://github.com/Azure/azure-storage-java/releases?after=v5.1.1) that:
```
- Removed deprecated startCopyFromBlob() on CloudBlob. Use startCopy() instead. 
```

Both Hadoop 2.7.3 and 2.8.3 invoke the startCopyFromBlob() method to write back to blob (so output Parquet files):
`java.lang.NoSuchMethodError: com.microsoft.azure.storage.blob.CloudBlob.startCopyFromBlob`

That prevents the use of any azure-storage library greater than 3.1.0 and we get back to the MD5 hash check errors. 3.1.0 was released in Sep 17, 2015 so it is getting up to 3 years old. I have also not been able to write back to blob with SAS tokens presumably due to the age of this library.

Basically I think the `hadoop-azure` library needs to be refactored to use the newer `azure-storage` libraries.

In the meantime disabling MD5 (per the initial post) will at least allow users to read and write to blob storage until the underlying issues can be addressed.

**Comment by jafreck** [comment]
I think we are slightly misunderstanding each other here. If you are on the latest versions of AZTK, Azure Storage functionality comes out of the box with aztk. The proper `hadoop-azure` jar is already placed on the Docker container. Same for the corresponding `azure-storage` jar for the version of hadoop that Spark is built with (`2.8.3` in our case). [You can reference maven to see the `azure-storage` dependency that `hadoop-azure` has.](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-azure/2.8.3)

The version of `hadoop-azure` is linked to the version of hadoop Spark is built with, so that should not be changed -- you should not add any additional `hadoop-azure` jars. Similarly, the version of `azure-storage` that is specified for hadoop-azure-2.8.3.jar is `azure-storage-2.2.0.jar`. So that too, should not be changed or updated.

Now, if you still experience this issue with the jars that come preloaded, you will have to raise the issue with the Azure storage team. On AZTK's side, we cannot just update the versions of `hadoop-azure` or `azure-storage` unless we update the version of Hadoop that Spark is built with.

**Comment by A139834** [comment]
Sorry if my comment above was a bit cryptic.

The summary is the issue does happen with the libraries included in `aztk` specifically `hadoop-azure-2.8.3.jar` and `azure-storage-2.2.0.jar`. So users will experience this issue with big reads of parquet files.

The rest of the comment was determining when/if subsequent versions of the `hadoop-azure` and `azure-storage` libraries fixed the issue. I found that `azure-storage-6+` fixes the read issue (but then prevents write) and will happen with both 2.7.x vs 2.8.x of `hadoop-azure`.

We can close this issue as it is an issue with `hadoop-azure`/`azure-storage` version interactions but at least this comment is indexed for when `aztk` users hit this issue.

**Comment by jafreck** [comment]
Ah I see, thanks for clarifying I misunderstood at first. 

Yes, this is unfortunate that `azure-storage-6+` is required. We can consider building Spark with a different Hadoop version target which has a `hadoop-azure` version with a dependency of `azure-storage-6+`.

Thanks for pointing this out, though, I'm sure it will come up for other users as well.
